solutions/platinum/bts-hotcold.mdx

```diff
@@ -0,0 +1,204 @@
+then this case must be handled slightly differently.
+
+**Case 2: $p$ is equal to either $a$ or $b$.**
+
+Let the first ancestor of $t$ that is on the path from $a$ to $b$
+be $l$. WLOG, let $b$ be equal to $p$. Then, we update the path from
+$a$ to $l$ and the path from $l$ to $b$.
+
+**Case 3: The LCA of $t$ with $a$ and $b$ is the same as $p$.**
+
+Like how we handled case 1, for this case we update the path from $a$ to
+$p$ and the path from $b$ to $p$.
+
+**Case 4: Either the LCA of $t$ and $a$ or the LCA of $t$ and $b$ lie on the path from $a$ to $b$.**
+
+In this case, split the path updates into 3 segments. Let the lower of the two LCAs mentioned
+be $l$, and the node which is an ancestor of $l$ be $a$. Then, just update the path from
+$a$ to $l$, $l$ to $p$, and the path from $b$ to $p$.
+
+### Handling Path Updates
+
+Note that based on how we did our casework on the paths, that all path updates are between a given
+node and its ancestor. Also, note that the path updates are just adding distances which either increase
+or decrease by a fixed amount every time. So, we can sort of do a difference array on a tree, where we DFS
+down the tree and apply the differences while walking back up the tree.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+class Tree {
+  private:
+	const int n;
+	const int log2dist;         // # of bits needed for binary lift
+	vector<vector<int>> &adj;   // reference to adjacency list
+	vector<vector<int>> lift;   // for binary lifting
+	vector<array<ll, 2>> diff;  // difference array updates
+	vector<int> depth;          // depth of each node
+
+	// these are for Euler tour
+	vector<int> tin;
+	vector<int> tout;
+	int timer = 0;
+
+	void tour(int u, int p) {
+		tin[u] = timer++;
+		lift[0][u] = p;
+		depth[u] = depth[p] + 1;
+		for (int i = 1; i < log2dist; i++) {
+			lift[i][u] = lift[i - 1][lift[i - 1][u]];
+		}
+		for (int v : adj[u]) {
+			if (v != p) { tour(v, u); }
+		}
+		tout[u] = timer - 1;
+	}
+
+	void dfs(int u, int p) {
+		for (int v : adj[u]) {
+			if (v == p) { continue; }
+			dfs(v, u);
+			diff[u][0] += diff[v][0] + diff[v][1];
+			diff[u][1] += diff[v][1];
+		}
+	}
+
+  public:
+	Tree(int n, vector<vector<int>> &adj)
+	    : n(n), log2dist((int)log2(n) + 1), adj(adj),
+	      lift(log2dist, vector<int>(n)), diff(n), depth(n), tin(n), tout(n) {
+		tin[0] = -1;
+		tout[0] = n + 1;  // ensures that LCA works
+		tour(1, 0);
+	}
+
+	bool is_ancestor(int u, int v) const {
+		return tin[u] <= tin[v] && tout[v] <= tout[u];
+	}
+
+	int lca(int u, int v) const {
+		if (is_ancestor(u, v)) { return u; }
+		if (is_ancestor(v, u)) { return v; }
+		for (int i = log2dist - 1; i >= 0; i--) {
+			if (!is_ancestor(lift[i][u], v)) { u = lift[i][u]; }
+		}
+		return lift[0][u];
+	}
+
+	int dist(int u, int v, int anc = -1) const {
```

Comment:
add a quick comment- "calculates the distance from u to anc and then to v (default is lca)"